### DISCUSSION TITLE
Extracted stiffness matrix not giving sensical modes

### DISCUSSION THREAD

**Discussion by jjalora** [sofa-framework/sofa#3134] [Develop in C++] [answered]
I've managed to extract the stiffness matrix _at the equilibrium point_ using the new (experimental) feature. 

I did a modal analysis of the Diamond soft robot (see [SoftRobot plugin](https://github.com/SofaDefrost/SoftRobots/tree/2a3ba021e0a046d9a71043a584211ca240d773b8/docs/sofapython3/tutorials/DiamondRobot)) by solving the generalized eigenvalue problem for the extracted $M$ and $K$ (in this case, I took -$K$ of what I get from SOFA) evaluated at the equilibrium point.

I get the following modes:
<img width="641" alt="image" src="https://user-images.githubusercontent.com/86177460/179246200-dd99da45-cb7b-4eb2-9132-5b71175ed3f6.png">

Clearly this is wrong. I'm wondering if any of you (particularly, @hugtalbot or @alxbilger) could replicate these results? Do I need to preprocess the $K$ matrix in any way prior to doing the analysis? 

Thanks for the help!

**Comment by jjalora** [comment, marked as answer]
Seeing that the first coupled modes look like translation modes, is it possible that the extracted stiffness matrix is the unconstrained version (i.e., not constrained to the base)? If so, would it be possible to extract the constrained version of the stiffness matrix?

**Reply by alxbilger** [reply]
I confirm that the stiffness matrix is not constrained. It should be made more explicit.
Currently, there is no option to add the constraints or not. But since you are in a Python environment, you can manipulate the matrix manually. If you know the index of a fixed DOF, you can set the rows and columns (`3 * index`, `3 * index + 1` and `3 * index + 2`) of the matrix to zero, except the diagonal where it is 1.

Let us know if it leads to better results.

**Reply by jjalora** [reply]
@alxbilger - this seems to be the solution! Thanks for the help. 

Now I get the following modal displacements after a modal analysis of the $M$ and $K$ matrices.
![image](https://user-images.githubusercontent.com/86177460/179634281-d29cb361-ff2d-4e96-ba0e-58a19aca0afc.png)

**Reply by alxbilger** [reply]
I am glad it worked
